src/schema/FroidSchema.ts

```diff
@@ -0,0 +1,880 @@
+      `Federation version must be a valid '${FED2_VERSION_PREFIX}x' version. Examples: v2.0, v2.3`
+    );
+
+    this.typeExceptions = options?.typeExceptions ?? [];
+    this.keySorter = options?.keySorter ?? defaultKeySorter;
+    this.nodeQualifier = options?.nodeQualifier ?? defaultNodeQualifier;
+    this.contractTags =
+      options?.contractTags
+        ?.sort()
+        .map<ConstDirectiveNode>((tag) =>
+          FroidSchema.createTagDirective(tag)
+        ) || [];
+
+    const currentSchemaMap = new Map<string, string>(schemas);
+    // Must remove self from map of subgraphs before regenerating schema
+    currentSchemaMap.delete(name);
+
+    // convert to a flat map of document nodes
+    const subgraphs = [...currentSchemaMap.values()].map((sdl) =>
+      FroidSchema.parseSchema(sdl)
+    );
+
+    // extract all definition nodes for federated schema
+    const allDefinitionNodes = subgraphs.reduce<DefinitionNode[]>(
+      (accumulator, value) => accumulator.concat(value.definitions),
+      []
+    );
+
+    this.filteredDefinitionNodes =
+      FroidSchema.removeInterfaceObjects(allDefinitionNodes);
+
+    this.extensionAndDefinitionNodes = this.getNonRootObjectTypes();
+    this.objectTypes = this.getObjectDefinitions();
+
+    this.findFroidObjectTypes();
+    this.generateFroidDependencies();
+
+    // build schema
+    this.froidAst = {
+      kind: Kind.DOCUMENT,
+      definitions: [
+        this.createLinkSchemaExtension(),
+        ...this.createCustomReturnTypes(),
+        this.createQueryDefinition(),
+        this.createNodeInterface(),
+        ...this.createObjectTypesAST(),
+      ],
+    } as DocumentNode;
+  }
+
+  /**
+   * Creates the AST for the object types that should be included in the FROID schema.
+   *
+   * @returns {ObjectTypeDefinitionNode[]} The generated object types.
+   */
+  private createObjectTypesAST(): ObjectTypeDefinitionNode[] {
+    return Object.values(this.froidObjectTypes).map(
+      ({node, finalKey, selectedKeyFields, selectedNonKeyFields}) => {
+        let froidFields: FieldDefinitionNode[] = [];
+        let externalFieldDirectives: ConstDirectiveNode[] = [];
+        const shareableFieldDirectives: ConstDirectiveNode[] = [
+          SHAREABLE_DIRECTIVE_AST,
+        ];
+        let froidInterfaces: NamedTypeNode[] = [];
+        if (FroidSchema.isEntity(node)) {
+          froidFields = [
+            FroidSchema.createIdField(this.getTagDirectivesForIdField(node)),
+          ];
+          externalFieldDirectives = [EXTERNAL_DIRECTIVE_AST];
+          froidInterfaces = [implementsNodeInterface];
+        }
+        const fields = [
+          ...froidFields,
+          ...selectedKeyFields.map((field) => ({...field, directives: []})),
+          ...selectedNonKeyFields.map((field) => ({
+            ...field,
+            directives: FroidSchema.isShareable(field)
+              ? // @todo test the shareable branch of this logic
```

Comment:
I don't know if I fully understand the @shareable here. I know that we _have_ to apply @shareable to a field if it's applied in one place, but how do we guarantee that router doesn't expect Froid to resolve the shareable field?

Reply:
My expectation is that we _would_ resolve the shareable field, but then it would have to be in the key. It would probably be better to just keep in @external which I think is compatible with @shareable fields (I'll verify).

Reply:
After reviewing this use case, you're right. `@shareable` will result in FROID schema being responsible for returning the non-key field. That being the case, it should always treat non-key fields (shareable or not) as `@external`. I'll make that change.